### PR TITLE
Reopen recently opened/closed files and modal navigation

### PR DESCRIPTION
**Pull request by krassowski**
## References

- Closes https://github.com/jupyterlab/jupyterlab/issues/5924
- Closes https://github.com/jupyterlab/jupyterlab/issues/15293
- Closes https://github.com/jupyterlab/jupyterlab/issues/11613
- Upstream discussion https://github.com/jupyterlab-contrib/jupyterlab-recents/issues/10

Punchlist
- [x] galata tests
- [x] add an icon [↗] to sidebar

## Code changes

Compared to `jupyterlab-recents` extension:

<details>

- remove polling for validation of paths; this was generating many background requests and was deemed undesirable overhead if we were to increase the number of requests by a factor of 10 (even though these were lightweight GET requests)
- do not send a get request to fetch the model/content type of the document for the that the user switch to - we already have it available at hand now thanks to integration directly into docmanager (this is not public/not emitted by the signal which is why the extension had to have this sub-optimal implementation)
- path validation is now conducted when opening the sub-menu, and when clicking on the file, limiting any overhead to the users of the feature
- when an item is discovered to be invalid after submenu was opened it will not be removed but greyed out to avoid jitter and user accidentally clicking a wrong item
- split recents into "recently opened" and "recently closed"
- use `IFileBrowserCommands` to explicitly depend on the availability of the filebrowser to open directories (and to fallback to document manager command if filebrowser does not provide such token

Because rather than adding a new package the extension gets integrated on the lower level by adding separate plugins to relevant existing packages (document manager, main menu) merging git history did no make much sense; instead the autorship was preserved by using `Co-authored-by` feature.

</details>

### Open Recent submenu

image

### Recently closed in the sidebar

image

Also seen is the [↗] button to open modal

### Reopen modal

reopen-widget-2

## Backwards-incompatible changes

None.

> File browsers are encouraged to implement `IFileBrowserCommands` which was previously a no-op; if they don't then the recents menu will not show the directories.

### CODE REVIEW

**Comment by krassowski** [comment]
Any suggestions on alternative shortcuts we could use?

> A part of me wishes that we had two sets of shortcuts: one for JupyterLab Desktop and one for when running in a standard browser.

**Comment by krassowski** [comment]
For now switching to <kbd>Ctrl</kbd> + <kbd>Alt</kbd> + <kbd>A</kbd> as it is not used by major browsers/operating systems as per https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts

**Comment by krassowski** [comment]
bot please update snapshots

**Comment by krassowski** [comment]
bot please update documentation snapshots

File: packages/docmanager-extension/schema/plugin.json [hunk 1]
```diff
@@ -149,6 +149,13 @@
       "title": "Rename Untitled File On First Save",
       "description": "Whether to prompt to rename untitled file on first manual save.",
       "default": true
+    },
+    "maximalRecents": {
```
Comment (jtpio) [inline packages/docmanager-extension/schema/plugin.json:153]: Another suggestion for the name of this setting:

```suggestion
    "maxNumberRecents": {
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks a lot for this @krassowski 

I have some minor suggestions.

File: packages/docmanager/src/recents.ts [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    );
+    return invalidPathsOrNulls.filter(x => typeof x === 'string') as string[];
+  }
+  /**
+   * Save the recent items to the state.
+   */
+  protected saveRecents(): void {
+    clearTimeout(this._saveRoutine);
+    // Save _recents 500 ms after the last time saveRecents has been called
+    this._saveRoutine = setTimeout(async () => {
```
[the remark this answers is not in the document]
Comment (krassowski) [reply packages/docmanager/src/recents.ts]: Good catch, done in 63c3e30fa79d0d327267dd164cb47fb8ea2f811d

File: packages/docmanager/src/recents.ts [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+import { IStateDB } from '@jupyterlab/statedb';
+import { IDisposable } from '@lumino/disposable';
+import { Debouncer } from '@lumino/polling';
+import { ISignal, Signal } from '@lumino/signaling';
+import { IRecentsManager, RecentDocument } from './tokens';
+
+/**
+ * Manager for recently opened and closed documents.
+ */
+export class RecentsManager implements IRecentsManager, IDisposable {
```
Comment (fcollonval) [inline packages/docmanager/src/recents.ts:17]: I guess this is not needed any longer, isn't it?

```suggestion
export class RecentsManager implements IRecentsManager {
```

**Review by fcollonval — APPROVED** [review]
Thanks @krassowski 

I have an optional suggestion.

File: packages/docmanager/src/recents.ts [hunk 4]
```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) Jupyter Development Team.
+ * Distributed under the terms of the Modified BSD License.
+ */
+
+import { PageConfig } from '@jupyterlab/coreutils';
+import { Contents, ServerConnection } from '@jupyterlab/services';
+import { IStateDB } from '@jupyterlab/statedb';
+import { IDisposable } from '@lumino/disposable';
```
Comment (krassowski) [inline packages/docmanager/src/recents.ts:9]:
```suggestion
```

**Comment by krassowski** [comment]
Snapshots will need updating after merging. I think one of the docs snapshots needs some work in general.

**Comment by krassowski** [comment]
but please update documentation snapshots

**Pull request merged**